probes/lttng-uprobes.c

```diff
@@ -0,0 +1,189 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <lttng-events.h>
+#include <lttng-tracer.h>
+#include <probes/lttng-probe-utils.h>
+#include <wrapper/irqflags.h>
+#include <wrapper/ringbuffer/frontend_types.h>
+#include <wrapper/uprobes.h>
+#include <wrapper/vmalloc.h>
+
+static
+int lttng_uprobes_handler_pre(struct uprobe_consumer *uc, struct pt_regs *regs)
+{
+	struct lttng_event *event =
+		container_of(uc, struct lttng_event, u.uprobe.up_consumer);
+	struct lttng_probe_ctx lttng_probe_ctx = {
+		.event = event,
+		.interruptible = !lttng_regs_irqs_disabled(regs),
+	};
+	struct lttng_channel *chan = event->chan;
+	struct lib_ring_buffer_ctx ctx;
+	int ret;
+
+	struct {
+		unsigned long ip;
+	} payload;
+
+	if (unlikely(!ACCESS_ONCE(chan->session->active)))
+		return 0;
+	if (unlikely(!ACCESS_ONCE(chan->enabled)))
+		return 0;
+	if (unlikely(!ACCESS_ONCE(event->enabled)))
+		return 0;
+
+	lib_ring_buffer_ctx_init(&ctx, chan->chan, &lttng_probe_ctx,
+				 sizeof(payload), lttng_alignof(payload), -1);
+
+	ret = chan->ops->event_reserve(&ctx, event->id);
+	if (ret < 0)
+		return 0;
+
+	/* event payload */
+	payload.ip = regs->ip;
+
+	lib_ring_buffer_align_ctx(&ctx, lttng_alignof(payload));
+	chan->ops->event_write(&ctx, &payload, sizeof(payload));
+	chan->ops->event_commit(&ctx);
+	return 0;
+}
+
+/*
+ * Create event description
+ */
+static
+int lttng_create_uprobe_event(const char *name, struct lttng_event *event)
+{
+	struct lttng_event_desc *desc;
+	struct lttng_event_field *fields;
+	int ret;
+
+	desc = kzalloc(sizeof(*event->desc), GFP_KERNEL);
+	if (!desc)
+		return -ENOMEM;
+	desc->name = kstrdup(name, GFP_KERNEL);
+	if (!desc->name) {
+		ret = -ENOMEM;
+		goto error_str;
+	}
+
+	desc->nr_fields = 1;
+	desc->fields = fields =
+		kzalloc(1 * sizeof(struct lttng_event_field), GFP_KERNEL);
+
+	if (!desc->fields) {
+		ret = -ENOMEM;
+		goto error_fields;
+	}
+	fields[0].name = "ip";
+	fields[0].type.atype = atype_integer;
+	fields[0].type.u.basic.integer.size = sizeof(unsigned long) * CHAR_BIT;
+	fields[0].type.u.basic.integer.alignment = lttng_alignof(unsigned long) * CHAR_BIT;
+	fields[0].type.u.basic.integer.signedness = lttng_is_signed_type(unsigned long);
+	fields[0].type.u.basic.integer.reverse_byte_order = 0;
+	fields[0].type.u.basic.integer.base = 16;
+	fields[0].type.u.basic.integer.encoding = lttng_encode_none;
+
+	desc->owner = THIS_MODULE;
+	event->desc = desc;
+
+	return 0;
+
+error_fields:
+	kfree(desc->name);
+error_str:
+	kfree(desc);
+	return ret;
+}
+
+int lttng_uprobes_register(const char *name,
+			   int fd,
+			   uint64_t offset,
+			   struct lttng_event *event)
+{
+	int ret;
+	struct inode *inode;
+
+	ret = lttng_create_uprobe_event(name, event);
+	if (ret)
+		goto error;
+
+	inode = lttng_get_inode_from_fd(fd);
+	if (!inode) {
+		printk(KERN_WARNING "Cannot get inode from fd\n");
+		ret = -EBADF;
+		goto error;
```

Comment:
I think we leak event->desc->name and event->desc on error here.

Reply:
Done.